Move the CUDA implementation of round to ATen.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25592 Simplify operator `sign` using the helper.
* #25423 Move the CUDA implementation of trunc to ATen.
* #25372 Move the CUDA implementation of floor to ATen.
* #25285 Move the CUDA implementation of rsqrt to ATen.
* **#25041 Move the CUDA implementation of round to ATen.**

Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev committed:
Move the CUDA implementation of round to ATen.

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


The underlying function is also changed from std::nearbyint to
std::round, because we do not want to be accidentally affected by
std::fsetround.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


The underlying function is also changed from std::nearbyint to
std::round, because we do not want to be accidentally affected by
std::fsetround.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

VitalyFedyunin commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -40,6 +40,14 @@ void ceil_kernel_cuda(TensorIterator& iter) {
+void round_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "round_cuda", [&]() {
+    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+      return std::round(a);
```
See #17443 we had reasons to use `nearbyint`

xuhdev replied:
I see, I have made the change and add a comment.

VitalyFedyunin reviewed (changes requested):
Overall good, but please use `nearbyint`

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

VitalyFedyunin commented:
Can you please rebase, unrelated tests seems to be failing.

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev commented:
@VitalyFedyunin rebased

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev commented:
@VitalyFedyunin I've updated this stack in accordance to #25689

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)

xuhdev commented:
@VitalyFedyunin I have resolved the CI failures. The reason is that HIP cannot recognize std::nearbyint. I resolved this issue by using wrappers.

xuhdev commented:
@VitalyFedyunin Is this ready for merge? It's been quite a while since this PR was made

xuhdev committed:
Update on "Move the CUDA implementation of round to ATen."


Fix #24617

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25041

Differential Revision: [D17114368](https://our.internmc.facebook.com/intern/diff/D17114368)
